### ISSUE TITLE
[feature] Allow binary compatibility between compilers

### ISSUE THREAD

**Issue by Loghorn** [conan-io/conan#17371]
### What is your suggestion?

I'd like if conan provided binary compatibility between different compilers.

To explain better what I mean, let's give you an example.

I'm on Windows and my project consists of a DLL and one EXE. I publish the DLL as a conan package and consume it from the EXE.

For reasons, I need to build the DLL with clang-cl and the EXE with msvc[^1] (I am using the clang coming with Visual Studio, so I'm sure the two output are ABI compatible, but even if they were not, the DLL exports only C functions, so there is no issue here).

The problem is that there I found no way to make conan download the binary created by a different compiler (re-building from sources is not an option and even if it was, it would use the wrong compiler).

I hoped the binary compatibility plugin would take care of it, but it only tries to find binary for different c++ standards, not different compilers.

[^1]: Actually, it's a bit more complicated: I need to build the DLL with both clang-cl and msvc, and be able to choose what version to use when installing the package in the EXE; otherwise, I could simply remove the compiler from the package id

### Have you read the CONTRIBUTING guide?

- [x] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @Loghorn 

Thanks for your question.

There are different ways to achieve what you want:

- You can directly define which dependencies use what compilers, using per-package settings. See for example https://docs.conan.io/2/reference/config_files/profiles.html#profile-patterns and https://docs.conan.io/2/reference/binary_model/extending.html#settings-vs-options-vs-conf. So something like ``conan install -s "mydep/*:compiler=clang" -s compiler=msvc`` is good to use msvc for the consumer app and other packages and clang for ``mydep``.
- You can customize the ``compatibility.py``, it is a plugin. By default it implements the compatibility accross ``compiler.cppstd``, but if you want a generic compatibility with fallbacks between clang and msvc (in either direction or even both), you can define that in the ``compatibility.py`` plugin  (this plugins is distributed and shared with ``conan config install/install-pkg`` together with the rest of user configuration). See https://docs.conan.io/2/reference/extensions/binary_compatibility.html

Please let us know if this helps.

**Comment by Loghorn** [comment]
Hi @memsharded 

Thank you for your reply.

The first option is problematic, because changing the compiler removes all the sub-options (`compiler.version`, `compiler.runtime`, etc.); of course I can specify each one of them, but it is tedious and error prone.

The second one could work, but the compatibility feature is still preview/experimental (not a big deal, unless you radically change it) and it means that for every new release of a complier I'd need to change it (I could do it, but I'd very much like if it was handled by conan itself, like the default `settings.yml`)

**Comment by memsharded** [comment]
> The first option is problematic, because changing the compiler removes all the sub-options (compiler.version, compiler.runtime, etc.); of course I can specify each one of them, but it is tedious and error prone.

Not really, this is done once in profile files, quite explicit and easy to test and maintain. Works very well in practice, we have seen tons of users happily doing this without issues.

> The second one could work, but the compatibility feature is still preview/experimental (not a big deal, unless you radically change it) and it means that for every new release of a complier I'd need to change it (I could do it, but I'd very much like if it was handled by conan itself, like the default settings.yml)

No plans to radically change it, it is already quite widely adopted, it is very unlikely that it will change. Same for the default ``compatibility.py`` current implementation, there are no plans at the moment to change it.

Still note that a ``compatibility.py`` custom compatibility still needs to define the version and other parameters mapping. It is not enough to just want to fallback from ``msvc`` to ``clang`` of viceversa, you need to define which versions map to which versions, etc. So basically, same scenario as per-package settings, when changing ``compiler``, it is necessary to define also its sub-settings,  this information cannot be automatically deduced.

**Comment by Loghorn** [comment]
I experimented a bit with `compatibility.py` but it has one big problem: it only gets called if there is no binary matching the current settings, but since I'm publishing the DLL with both clang and msvc, when building the exe with msvc it will always find a match, so the plugin (or the function in the conanfile) is never invoked

**Comment by memsharded** [comment]
> I experimented a bit with compatibility.py but it has one big problem: it only gets called if there is no binary matching the current settings, but since I'm publishing the DLL with both clang and msvc, when building the exe with msvc it will always find a match, so the plugin (or the function in the conanfile) is never invoked

I am afraid that I don't understand the issue or feature request then. If you compiled the package with the 2 different compilers, then no ``compatibility.py`` is needed, that is totally expected. Then the problem reduces to the user deciding which one to use, and the way to do this is with the per-package settings, just defining which compiler to use with ``-s mypkg/*:compiler=clang ...``.

**Comment by Loghorn** [comment]
As I said, just defining the compiler won't work, you'll need to define `compiler`, `compiler.version`, `compiler.runtime`, `compiler.runtime_version`, `compiler.runtime_type`, and maybe some more I am forgetting now

**Comment by memsharded** [comment]
> As I said, just defining the compiler won't work, you'll need to define compiler, compiler.version, compiler.runtime, compiler.runtime_version, compiler.runtime_type, and maybe some more I am forgetting now

Sure, how would the system guess which sub-settings you want if not specified?

**Comment by Loghorn** [comment]
By trying all the valid combination of settings for that particular compiler?
But at least it would be nice if settings that are common between the two compiler (e.g. complier.runtime) would not be obliterated by just setting the compiler

**Comment by memsharded** [comment]
> By trying all the valid combination of settings for that particular compiler?

No, you pick the one that you want to use. In the same way you say: "for this build I don't want to use ``compiler=msvc``, but ``compiler=clang``", then you also say, "and instead of the ``compiler.version=193``, I want the ``compiler.version=17``. No try all possible versions, this is a user decision, you decide which version you want to use.

For other arguments there are some other alternatives. For example, you can use jinja syntax to keep settings consistent:
```
[settings]
{% set runtime = "dynamic" %}
os=Windows
arch=x86_64
compiler=msvc
compiler.version=190
compiler.runtime={{runtime}}

zlib/*:compiler=clang
zlib/*:compiler.runtime={{runtime}}
```

Another alternative would be to use the ``profile.py`` plugin to auto-complete some of the missing sub-settings, based on your own custom logic.

In any case, we have already discussed this, and Conan will not try to implement any other binary compatibility rule across compilers. Binary compatibility is not only challenging, but there is also no silver bullet, and different users want different binary compatibility rules. We introduced the ``compiler.cppstd`` rules in the default ``compatibility.py`` after years of running at scale in ConanCenter, with quite good experience that it holds in the majority of cases (it also has exceptions, this is why there is an ``extension_properties`` attribute to allow defining such exceptions. This same experience tells us that there is no possible default binary compatibility across compilers that Conan could define that would be better than the current strategy of not defining any. This is why Conan provides the necessary tools to implement your own logic regarding the compatibility, but it will no provide such compatibility.

**Comment by Loghorn** [comment]
I see, thank you for your answer, I'll figure it out :)
Btw, is there a way to access the content of `settings.yml` file from `conanfile.py`? (other than opening and parsing it myself, of course)

**Comment by memsharded** [comment]
Not the full contents of ``settings.yml``, but there is the ``self.settings.possible_values()`` can list the different possible settings values for a given setting or subsetting, like ``compiler_versions = self.settings.compiler.version.possible_values()``.

This would be something quite extreme and exceptional, in general the recommended approach in ``conanfile.py`` recipes is just to read the values coming from profiles. In any case the ``settings`` are **read-only** in recipes, their value cannot be changed, check https://docs.conan.io/2/knowledge/guidelines.html:

> Settings and configuration (conf) are read-only in recipes: The settings and configuration cannot be defined or assigned values in recipes. Something like self.settings.compiler = "gcc" in recipes shouldn’t be done. That is undefined behavior and can crash at any time, or just be ignored. Settings and configuration can only be defined in profiles, in command line arguments or in the profile.py plugin.

**Comment by Loghorn** [comment]
I wasn't planning to change the settings, I was planning to enumerate all the settings in the `compatibility` function inside the recipe to generate all possible combinations ;-)
Anyway, I'm now using the compatibility function to make a different compiler compatible, but I've found a problem :-(
I have a static library that I build with clang, a shared library that uses the static library and is compiled with clang, and finally an exe that is built with msvc. Using compatibility I can "change" the settings when installing the package in the exe and it works, but the `requires` for the static library is using the id for the msvc version so there is no match.

This is what I have in the conan cache/remote (it's a bit redacted, but the relevant info are there):
```
            8bcddd54f1d222ed7a619c20ca78af86c18edd5e
              info
                settings
                  arch: x86_64
                  build_type: Release
                  compiler: clang
                  compiler.cppstd: 20
                  compiler.runtime: dynamic
                  compiler.runtime_type: Release
                  compiler.runtime_version: v143
                  compiler.version: 17
                  os: Windows
                full_settings
                requires
                  gcem/1.17.0#0fc2e1bd4a3f19487e54948541256a67:da39a3ee5e6b4b0d3255bfef95601890afd80709
                  staticlib/3.0.0.0#35c60c6999e3566ce48c8f7c1f08fb0ab7bb6208:43498b78e88d6983d0f2a903e741a0263638a9aa
                full_requires
                options
                  shared: True
                full_options
                recipe_hash:
                environment
 ```
And this is what conan is looking for (I added a print in `graph\compatibility.py` just after it gets the pid of the compatible package:
```
[settings]
arch=x86_64
build_type=Release
compiler=clang
compiler.cppstd=20
compiler.runtime=dynamic
compiler.runtime_type=Release
compiler.runtime_version=v143
compiler.version=17
os=Windows
[options]
shared=True
[requires]
gcem/1.17.0#0fc2e1bd4a3f19487e54948541256a67:da39a3ee5e6b4b0d3255bfef95601890afd80709
staticlib/3.0.0.0#35c60c6999e3566ce48c8f7c1f08fb0ab7bb6208:b360b6000eff37ef0c14f880bd7062648b9a2944
```
The `gcem` library is header-only, so the package_id always matches, but the `staticlib` ids are different: the one in the remote is the one generated when building with clang, the other one is the one generated when building with msvc.

I noticed that for some binaries, the requirement are not "complete" but they are like `libname/1.2.Z` without any hash, but it's not clear to me why some are like this and some aren't, even if the binary is generated from the same source/recipe.

**Comment by memsharded** [comment]
> Anyway, I'm now using the compatibility function to make a different compiler compatible, but I've found a problem :-(
I have a static library that I build with clang, a shared library that uses the static library and is compiled with clang, and finally an exe that is built with msvc. Using compatibility I can "change" the settings when installing the package in the exe and it works, but the requires for the static library is using the id for the msvc version so there is no match.

Yes, I can see how this could be a challenge:
- There is one binary copy of the library built with clang, that is linked inside the shared library
- There another, different binary copy of the library inside the executable.

Still, by default, the dependency graph will be as:
```
exe -> shared_lib -> static_lib
 \-----------------------/
```
That is a half-diamond structure, and by default both the ``exe`` and the ``shared_lib`` should depend on the exact same thing, including the same version, otherwise it is a dependency version conflict. Even if there are 2 copies of the static_lib binary code at linkage time, because there are 2 linkages happening there, the dependency graph by default requires to have 1 dependency.

The dependency from the ``shared_lib`` to the ``static_lib`` could be made private with ``self.requires("static_lib/version", visible=False)`` which would conver the dependency graph to:

```
exe -> shared_lib -> static_lib
 \------------------>static_lib
```
With this approach, it is possible to depend on different versions of the ``static_lib``. It is of course mandatory that the ``shared_lib`` fully isolates the ``static_lib`` linkage, making it fully both ABI and API private, otherwise there will be linkage errors while linking the ``exe``.

Still, no matter what is defined in ``compatibility()`` those are fallbacks for the existence of a given binary. You can define that Conan will look for a library package built with clang, and if it doesn't exist, then use the binary built with msvc. Or viceversa. But it is not possible to define to use different compilers for the exact same package that appears multiple times in the dependency graph, because the dependency. As the ``static_lib`` name is the same, it is not possible to differentiate it in the ``compatibility`` or even in the profiles per-package settings as ``static_lib/*:compiler=clang``.

So maybe for this case, the best option could be to have the ``shared_lib`` to fully vendor the ``static_lib`` with the ``vendor=True`` feature, check: https://blog.conan.io/2024/07/09/Introducing-vendoring-packages.html. This is intended mostly for other release cases, but it might be a possibility to handle this scenario.

> I noticed that for some binaries, the requirement are not "complete" but they are like libname/1.2.Z without any hash, but it's not clear to me why some are like this and some aren't, even if the binary is generated from the same source/recipe.

This depends on the linkage model, called ``embed`` vs ``non-embed`` in Conan ``package_id`` terms, check: https://docs.conan.io/2/reference/binary_model/dependencies.html. you will see there the explanations about the ``libname/1.2.Z``

**Comment by rickrick318** [comment]
hi @memsharded, does these 2 ways work for conan 1.59.0? if not, is their any other way to do this on conan v1? thanks

> Hi @Loghorn
> 
> Thanks for your question.
> 
> There are different ways to achieve what you want:
> 
> * You can directly define which dependencies use what compilers, using per-package settings. See for example https://docs.conan.io/2/reference/config_files/profiles.html#profile-patterns and https://docs.conan.io/2/reference/binary_model/extending.html#settings-vs-options-vs-conf. So something like `conan install -s "mydep/*:compiler=clang" -s compiler=msvc` is good to use msvc for the consumer app and other packages and clang for `mydep`.
> * You can customize the `compatibility.py`, it is a plugin. By default it implements the compatibility accross `compiler.cppstd`, but if you want a generic compatibility with fallbacks between clang and msvc (in either direction or even both), you can define that in the `compatibility.py` plugin  (this plugins is distributed and shared with `conan config install/install-pkg` together with the rest of user configuration). See https://docs.conan.io/2/reference/extensions/binary_compatibility.html
> 
> Please let us know if this helps.

**Comment by memsharded** [comment]
Only the per-package settings exists in Conan 1, @rickrick318. The ``compatibility.py`` plugin is exclusive of Conan 2

**Comment by Loghorn** [comment]
> The dependency from the `shared_lib` to the `static_lib` could be made private with `self.requires("static_lib/version", visible=False)` which would conver the dependency graph to:
> 
> ```
> exe -> shared_lib -> static_lib
>  \------------------>static_lib
> ```
> 
> With this approach, it is possible to depend on different versions of the `static_lib`. It is of course mandatory that the `shared_lib` fully isolates the `static_lib` linkage, making it fully both ABI and API private, otherwise there will be linkage errors while linking the `exe`.

@memsharded, the `visible=False` works, but I don't understand why static_lib is still in the graph if the exe only requires shared_lib. Is there a way to really make the static_lib used by shared_lib invisible to the outside world?
My situation is like this: 
```
exe -> shared_lib -> static_lib_used_only_internally
 |          v
 \------->static_lib
```
I'm fine with the diamond introduced by static_lib, but I want to remove completely from the graph the static_lib_used_only_internally because it's in no way exposed and shared_lib has already been linked with it (it would be different if shared_lib was static, in that case I'd need to link with the static_lib_used_only_internally, of course)

**Comment by memsharded** [comment]
> @memsharded, the visible=False works, but I don't understand why static_lib is still in the graph if the exe only requires shared_lib. Is there a way to really make the static_lib used by shared_lib invisible to the outside world?
My situation is like this:

It is not fully invisible, because it must be known, as it determines the identity and ``package_id`` of the shared library. If there is a new recipe revision or new package-id of the static-lib, that  means that a new binary for ``shared-lib`` is necessary. This is only possible to compute if the graph is expanded to include ``static-lib``.

In practice, most times the ``static-lib`` will be marked as "Skip", that means that only the recipe will be downloaded (very fast), which is the necessary part to compute the identity=package-id, but the actual binary (the heavy part that takes longer to download) will not be downloaded. So the ``static-library`` will not be included or linked at all by the consumers at build time, it only exists in its representation of the graph, not as a binary artifact to link.

There are other mechanisms like the ``vendor=True`` to fully isolate the shared-library, but that breaks the identity-package_id computation and is only recommended for actual distribution of packages outside of the organization in which the version at release time account for all the binary variability. Do not use this mechanism in regular Conan dependency graphs, as it will lead to broken package-id computation.

**Comment by memsharded** [comment]
Is there any further question here?

Brief summary:

- Conan does not provide a default compatibility between compilers
- Users can implement they own binary compatibility rules in ``compatibility.py``
- Shared libraries linking static libraries depend on the full package binary "identity", meaning that a shared library with a dependency that is a static library that changes its compiler, even if the compiler is "binary" compatible, it will still change its ``package_id`` and require a rebuild. One thing is binary-compatibility and the other is identity, for example the binary with different compilers could have different performance, and that should affect the shared library as well.
- Users can also change the defaults ``package_id`` model and drop the dependency on the binary identity of the dependency.

If there are no further questions, I think the ticket can be closed as unplanned or as question responded.

**Comment by memsharded** [comment]
Closing as responded, thanks for the feedback!
